src/s3/multipartcopy.rs

```diff
@@ -0,0 +1,930 @@
+    }
+
+    /// Uploads a part of the source object to the destination as part of the multipart upload.
+    ///
+    /// # Arguments
+    ///
+    /// * `part` - The `CopyUploadPart` containing the parameters for the upload.
+    ///
+    /// # Returns
+    ///
+    /// The completed part containing the ETag and part number.
+    #[instrument(skip(self))]
+    async fn upload_part_copy(
+        &self,
+        part: CopyUploadPart<'_>,
+    ) -> Result<CompletedPart, S3MultipartCopierError> {
+        let copy_source = S3MultipartCopier::copy_source(part.src);
+
+        let response = self
+            .client
+            .upload_part_copy()
+            .bucket(&part.dst.bucket)
+            .key(&part.dst.key)
+            .part_number(part.part_number)
+            .upload_id(part.upload_id)
+            .copy_source(copy_source)
+            .copy_source_range(part.byte_range.as_string())
+            .send()
+            .await?;
+
+        Ok(CompletedPart::builder()
+            .set_e_tag(
+                response
+                    .copy_part_result
+                    .ok_or(S3MultipartCopierError::MissingCopyPartResult)?
+                    .e_tag,
+            )
+            .part_number(part.part_number)
+            .build())
+    }
+
+    /// Completes the multipart upload by combining all parts.
+    ///
+    /// # Arguments
+    ///
+    /// * `upload_id` - The upload ID of the multipart upload.
+    /// * `parts` - A vector of completed parts.
+    ///
+    /// # Returns
+    ///
+    /// An empty result indicating success.
+    #[instrument(skip(self))]
+    async fn complete_multipart_upload(
+        &self,
+        upload_id: &str,
+        mut parts: Vec<CompletedPart>,
+    ) -> Result<(), SdkError<CompleteMultipartUploadError>> {
+        parts.sort_by_key(|part| part.part_number);
+        let completed_multipart_upload = CompletedMultipartUpload::builder()
+            .set_parts(Some(parts))
+            .build();
+
+        self.client
+            .complete_multipart_upload()
+            .bucket(&self.destination.bucket)
+            .key(&self.destination.key)
+            .upload_id(upload_id)
+            .multipart_upload(completed_multipart_upload)
+            .send()
+            .await?;
+
+        Ok(())
+    }
+
+    /// Performs a multipart copy of a large object from the source bucket to the destination bucket.
+    ///
+    /// # Returns
+    ///
+    /// An empty result indicating success.
+    #[instrument(skip(self))]
+    pub async fn send(&self) -> Result<(), S3MultipartCopierError> {
+        tracing::info!("Starting multipart copy");
+        let source_size =
+            get_source_size(&self.client, &self.source.bucket, &self.source.key).await?;
+
+        tracing::info!(
+            source_size = source_size.as_ref(),
+            part_size = self.part_size.as_ref(),
+        );
+
+        if source_size.as_ref() <= self.part_size.as_ref() {
```

Comment:
**observation:** There's a slight contraction here with line `488`. We use `source <= part` here, but `source < part` (but spelt backwards) at `488` 🤔 I don't know if it makes much difference which one we change. If the block size is **equal to** the part size, then I guess an atomic copy will be more efficient? In that case we should change `488` to be `source <= part`. I'd also flip line `488` so that the inequality is spelt in the same order as `466`, just to make it easier to see how these two lines relate to each other.